### ISSUE TITLE
Inline rename: symbol duplicated when a generic type parameter

### ISSUE THREAD

**Issue by cartermp** [Microsoft/visualfsharp#4231]
```fsharp

type Record = { IntVal: int; StringVal: string }

[<AbstractClass>]
type AbstractFoo<'T>() =
    abstract member Foo: 'T -> 'T

type C() =
    inherit AbstractFoo<Record>()
    override __.Foo(r) = { r with IntVal = 12 }
```

Rename `Record` at either the definition or use site:

![rename-fail](https://user-images.githubusercontent.com/6309070/35188789-a7c7efcc-fdf1-11e7-9d5b-4d6545913694.gif)

Info:

- .NET Framework 4.6.1 project
- VS 15.6 Update 2
- VF# tools 15.4.1.17112103

**Comment by cartermp** [comment]
Reduced repo:

```fsharp
[<AbstractClass>]
type AbstractFoo<'T>() =
    abstract member Foo: 'T -> 'T

type C() =
    inherit AbstractFoo<int>()
    override __.Foo(i) = i
```

0. Ensure you're debugging and set a breakpoint here: https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/NameResolution.fs#L1465
1. Click `int` in the above code.

**Expected:**

Single symbol use for `int`, which should be a `UseInType` Occurage.

**Actual:**

Three symbol uses for `int`:

1. `UseInType`:
![image](https://user-images.githubusercontent.com/6309070/36358450-6341e964-14c3-11e8-94ed-a52dfcec8145.png)

2. `UseInType`:
![image](https://user-images.githubusercontent.com/6309070/36358463-9f7cc426-14c3-11e8-8721-cf04d969e2e1.png)

3. `Use`:
![image](https://user-images.githubusercontent.com/6309070/36358471-c1381552-14c3-11e8-8385-b8987a373e17.png)

This surfaces in inline rename in the original issue text because the code there just finds the spans for each symbol use's location and replaces the text. Since the inline rename service finds multiple symbols for the same thing, it has multiple spans to apply the change to, hence the bug.

@dsyme The [capturedNameResolutions](https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/NameResolution.fs#L1458) that are used to find symbols are filled when the document is typechecked - as far as I can tell, code to find the symbols has no bug in it.  **Is it correct behavior to find what amounts to duplicate symbols for the same thing?** If not, then this might actually be a bug in the typechecker.

**Comment by dsyme** [comment]
> @dsyme The capturedNameResolutions that are used to find symbols are filled when the document is typechecked - as far as I can tell, code to find the symbols has no bug in it. Is it correct behavior to find what amounts to duplicate symbols for the same thing? If not, then this might actually be a bug in the typechecker.

There can be duplicates in this list - there are some phases of checking we run multiple types, especially related to type parameter declarations, and it's hard to prevent duplicates creeping in.  IIRC there is code later to remove duplicates - it's likely that this isn't dealing with type parameters

**Comment by cartermp** [comment]
Makes sense. Yes, currently there doesn't seem to be anything in FCS which will filter out "duplicate" symbols (either that or I haven't looked hard enough). There is code in Inline Rename which will distinct on symbols, but because in this case they are distinct symbols, they're not filtered out.

**Comment by dsyme** [comment]
I think I was thinking of RemoveDuplicateItems  in service.fs

**Comment by dsyme** [comment]
I think I was thinking of RemoveDuplicateItems in service.fs. But that's for autocomplete, not symbol usage.

There's ItemsAreEffectivelyEqual in NameResolution.fs, which is made public via IsEffectivelySameAs.  But we don't make GetAllUsesOfAllSymbols distinct by this.  I'm a bit unsure of the perf characteristics of `ItemsAreEffectivelyEqual` and GetAllUsesOfAllSymbols  is a bit of a hot path.  

Just adding code to selectively de-duplicate type parameters would probably work
